src/main/java/com/yahoo/memory/AllocateDirectMap.java

```diff
@@ -103,25 +107,35 @@ public void close() {
 
   //Does the actual mapping work
   @SuppressWarnings("resource")
-  static final ResourceState mapper(final ResourceState state) throws Exception {
+  static final ResourceState mapper(final ResourceState state) throws IOException {
     final long fileOffset = state.getFileOffset();
     final long capacity = state.getCapacity();
 
     final File file = state.getFile();
 
     if (isFileReadOnly(file)) {
-      state.setResourceReadOnly(); //The file itself could be writable
+      if (!state.isResourceReadOnly()) {
+        throw new IllegalStateException("Cannot map read-only file into writable memory.");
+      }
     }
 
-    final String mode = "rw"; //we can't map it unless we use rw mode
+    final String mode = state.isResourceReadOnly() ? "r" : "rw";
     final RandomAccessFile raf = new RandomAccessFile(file, mode);
     state.putRandomAccessFile(raf);
+    if (fileOffset + capacity > raf.length()) {
+      if (state.isResourceReadOnly()) {
+        throw new IllegalStateException(
+            "File is shorter than the region that is requested to be mapped: file length=" +
+                raf.length() + ", mapping offset=" + fileOffset + ", mapping size=" + capacity);
+      } else {
+        raf.setLength(fileOffset + capacity);
+      }
+    }
     final FileChannel fc = raf.getChannel();
-    final long nativeBaseOffset = map(fc, fileOffset, capacity);
+    final int mapMode = state.isResourceReadOnly() ? MAP_RO : MAP_RW;
+    final long nativeBaseOffset = map(fc, mapMode, fileOffset, capacity);
     state.putNativeBaseOffset(nativeBaseOffset);
 
-    // length can be set more than the file.length
-    raf.setLength(capacity);
     final MappedByteBuffer mbb = createDummyMbbInstance(nativeBaseOffset);
```

Comment:
@leerho why this is needed?

Reply:
What?  the comment? or the raf.setLength()?

Reply:
No, why dummy MBB is created.

Reply:
It is a performance optimization.  The native methods node0 and madvise and the dummy mbb are closely connected.  Praveen, who was the original author on this class discovered the underlying optimization by studying the native code, which also allocates a dummy mbb with the right capacity. Then by initially setting only one byte per page from the file speeds up the os page loading process considerably.  Praveen actually ran performance tests both with and without this optimization and it makes a huge difference in file load time.

Reply:
I didn't understand this much from this explanation. As far as I can understand, MBB is created in order to be able to call `load0()`, `isLoaded0()` and `force0()` without defining them via JNI extensions on your own, rather than perform performance optimization per se.

However, there is a problem - when DirectByteBuffer constructor is called, a separate `Cleaner` (independent from the one created by Memory library itself) is created, that could lead to freeing the same memory two times.

Instead of creating a `DirectByteBuffer` (which is always a `MappedByteBuffer` - note that `DirectByteBuffer` extends `MappedByteBuffer`, that is a little odd) this way, this code should rather use facility similar to `BaseWritableMemoryImpl.wrap()`, duplicating the same `ZERO_DIRECT_BUFFER`, allocated in that class, and reinstalling address in it. Or just use `ZERO_DIRECT_BUFFER` directly, because `load0()`, `isLoaded0()` and `force0()` methods are effectively static, as far as I can see.

Reply:
Another problem is that those `Method` objects are not cached in static final constants, and obtained and created on each call.

Reply:
Could you please fix those issues?

Reply:
> Instead of creating a DirectByteBuffer (which is always a MappedByteBuffer - note that DirectByteBuffer extends MappedByteBuffer, that is a little odd) this way, this code should rather use facility similar to BaseWritableMemoryImpl.wrap(), duplicating the same ZERO_DIRECT_BUFFER, allocated in that class, and reinstalling address in it. Or just use ZERO_DIRECT_BUFFER directly, because load0(), isLoaded0() and force0() methods are effectively static, as far as I can see.

If we did this then we would have duplicate Cleaners!

Reply:
Hm, why? `DirectByteBuffer.duplicate()` doesn't duplicate the associated Cleaner object. And in fact, we don't need to duplicate ZERO_DIRECT_BUFFER, we could use it right away, because `load0()`, `isLoaded0()` and `force0()` are effectively static.

Reply:
> DirectByteBuffer.duplicate() doesn't duplicate the associated Cleaner object. 

Because the DBB constructor that is used by `duplicate()` and `slice()` sets `cleaner = null`.

Note: Our current `ZERO_DIRECT_BUFFER` actually allocates memory of 1 byte off-heap and has a valid `Cleaner` with a `Runnable` unmapper.  If a system is pageAligned, it would allocate a minimum of 1 page (4KB, I think).  

> And in fact, we don't need to duplicate ZERO_DIRECT_BUFFER, we could use it right away, because load0(), isLoaded0() and force0() are effectively static.

This might lead to trouble.  The original address of the ZDB will have been overwritten.  When it comes time to actually deallocate the ZDB it will be pointing to an address that has already been deallocated.  The ZDB will attempt to deallocate again calling `unsafe.freeMemory(...)` and `Bits.unreserveMemory(...)`.  I think the unsafe call can throw an error and the Bits fields will be improperly decremented (even if it is only by 1). Meanwhile the original allocation of the ZDB, although tiny, is a leak.

What we really need is a shell (zero) MBB/DBB that has no side effects: no resource allocation (heap or off-heap), and no Cleaner( so no entry into the PhantomReference, and no Runnable unmapper).  Then we swap in the address and length for our buffer use cases.  For the mapping side we call the load0, etc, methods, and for the WritableByteChannel we pass it to the WBC as a ByteBuffer, which it is. In both cases we know the memory we are passing in is already allocated. Whenever this shell-new-MBB goes out of scope, it has no resources and does nothing

The way we are creating it in AllocateDirectMap it does create an entry into the `Cleaner's PhantomReference` table (even though the unmapper is null)...this isn't super clean either, but probably does no harm.  Need to think about this.

Reply:
> The original address of the ZDB will have been overwritten. 

Could you please point to the place in code where it is going to be overwritten?

Reply:
If we used the same ZDB from BaseWritableMemoryImpl in AllocateDirectMap [here](https://github.com/DataSketches/memory/blob/master/src/main/java/com/yahoo/memory/AllocateDirectMap.java#L193):

> And in fact, we don't need to duplicate ZERO_DIRECT_BUFFER, we could use it right away

Then: the address would get overwritten like this [here](https://github.com/DataSketches/memory/blob/master/src/main/java/com/yahoo/memory/BaseWritableMemoryImpl.java#L375).

In conclusion we always must duplicate. (it is pretty low cost).

Reply:
I didn't suggest to call anything similar to wrap(). Address and size are obtained from ResourceState. Then one of the native methods is called provided ZDB (*without modifying it at all*), address, and size.

Reply:
Sorry, I misunderstood.  AllocateDirectMap now uses ZERO_DIRECT_BUFFER as you suggested.   I also moved ZERO_DIRECT_BUFFER to `AccessByteBuffer` as it is now used in multiple places.  I'd like to keep as much of the ByteBuffer specialized code in the ABB class.

I moved the "map(long, int)" function in `BaseWritableMemoryImpl` to `AccessByteBuffer`  and renamed it "getDummyDirectByteBuffer" although it is returned as a ByteBuffer because DirectByteBuffer is not public.

Reply:
Ok